initial draft allowing takers to optionally complete transactions wit…

AdamISZ opened the pull request:
…h less than the initially requested number of makers; sweep function is unchanged

It occurred to me that this is a much simpler, faster, and possibly more robust way of dealing with unresponsive makers who don't even provide utxos or signature up to !ioauth.

This is not really intended to supercede the idea of more intelligent selection, something like outlined [here](https://gist.github.com/AdamISZ/79457f1c20bc0702d299d873c899676a) but is a more basic idea that is simpler.
 
If the number of full respondants in the `!fill` - `!ioauth` phase is less than that specified (by -N or similar), but is at least equal to `jm_single().config.getint("POLICY", "minimum_makers")`, then continue anyway - the change will be calculated correctly, higher than originally anticipated, and the txfee will be less than originally anticipated but it is already re-calculated at this point.

The timeout code in `recover_from_nonrespondants` thus checks the above arithmetic, if sufficient, it calls `recv_txio` with dummy values, which flags the code to skip the first step and pass straight into calculation of our change output and construction of `!tx` messages (nothing for the makers changes; their outputs are already set by their own fees).

Nothing changes for sweep; there was previously no way to handle any change in the input set of orders, and there still isn't (at least I haven't thought of one yet).

I've done various tests with randomly malicious makers (using `test/ygrunner.py` and inserted random failure-to-ioauth into `maker.py`) and it seems to work as intended.

User can flag not to use this feature by setting `POLICY.minimum_makers` to 0.

Review greatly welcomed. This needs careful testing, even if it is a fairly simple change (at least in comparison to other ways of dealing with the problem!)

AdamISZ commented:
Tried on main net, chose 2 known malicious out of 6 using -P, got a 4 party tx completed.

AlexCato commented on joinmarket/configure.py:
```diff
@@ -126,6 +127,11 @@ def jm_single():
+# to complete, accounting for the fact that some makers might not be
+# responsive. Should be an integer >=2 for privacy, or set to 0 if you
+# want to disallow any reduction from your chosen number of makers.
+minimum_makers = 2
```
I also believe this should be opt-out. "First user" experience is important, also default-settings should just work. Which they currently dont, due to these unresponsive bots.

AlexCato replied:
Recommendation for user though: at the point where they confirm that they will pay 0.xx % of cjfees in total, there should be a note if minimum_makers > 0. Something like:
If some makers do not respond, we will still create a coinjoin with at least (minimum_makers) makers. If you do not desire this behavior, you can set "minimum_makers = 0" in joinmarket.cfg.

AdamISZ replied:
@AlexCato I agree. Will push and squash that edit shortly.

AdamISZ commented:
Additional notes as of 4fe8e25 (will squash before merge, don't want to bork comments for now)

This commit does 3 things:
1. Trivial, remove `CoinJoinTx.actual_respondants` which was a holdover from working on another branch, it is completely redundant with the existing `active_orders.keys()` here.
2. I had removed the existing smart order-reselection on initial failure for the case `minimum_makers=0` which was unnecessary; that code is reinstated in the block `elif minmakers == 0` in `recover_from_nonrespondants()`. Thus, the code leaves functionality unchanged if people disable this option.
3. Added an informative message as per @AlexCato 's suggestion, in `sendpayment.py` so that if choose-from-subset is enabled, the user is warned that the maker count can go down.

I did a bunch of tests of cases: (1) minmakers 0, no malicious bots, (2) minmakers 0, randomly malicious bots, (3) minmakers 2, no malicious bots, (4) minmakers 2, malicious bots and (5) sweeps with malicious bots (minmaker setting not relevant).

The tests seem to behave as expected. I'll run a couple more tumbler tests today, this time including malicious behaviour in bots to see if it handles it OK.

AdamISZ commented:
Tumbler tests with inserted maker maliciousness seem to be working fine. I'd anticipate merging this in the next few days - and making a new release - unless anyone objects; additional tests by others hugely appreciated.

adlai commented on joinmarket/taker.py:
```diff
@@ -405,43 +412,59 @@ def self_sign_and_push(self):
+                self.msgchan.fill_orders(new_orders, self.cj_amount,
+                                         self.kp.hex_pk(), self.commitment)
+            else:
+                log.info("Two few makers responded to complete, trying again.")
```
s/Two/Too/
